actors/evm/src/interpreter/precompiles/mod.rs

```diff
@@ -14,79 +14,117 @@ use fvm::{call_actor, get_actor_type, get_randomness, lookup_delegated_address,
-        lookup_delegated_address,     // lookup_delegated_address 0x0b
-        get_actor_type,     // get actor type 0x0c
-        get_randomness,     // rand 0x0d
-        call_actor,         // call_actor 0x0e
+        resolve_address,            // 0xfe00..01 resolve_address
+        lookup_delegated_address,   // 0xfe00..02 lookup_delegated_address
+        get_actor_type,             // 0xfe00..03 get actor type
+        get_randomness,             // 0xfe00..04 rand
+        call_actor,                 // 0xfe00..05 call_actor
     }
 }
 
 pub struct Precompiles<RT>(PhantomData<RT>);
 
 impl<RT: Runtime> Precompiles<RT> {
-    const PRECOMPILES: [PrecompileFn<RT>; 14] = gen_precompiles();
-    const MAX_PRECOMPILE: U256 = {
+    const EVM_PRECOMPILES: [PrecompileFn<RT>; 9] = gen_evm_precompiles();
+    const MAX_EVM_PRECOMPILE: U256 = { U256::from_u64(Self::EVM_PRECOMPILES.len() as u64) };
+
+    const NATIVE_PRECOMPILES: [PrecompileFn<RT>; 5] = gen_native_precompiles();
+    /// 0x<leading zeros>FE..00
+    const NATIVE_PRECOMPILE_START: U256 = {
         let mut limbs = [0u64; 4];
-        limbs[0] = Self::PRECOMPILES.len() as u64;
+
+        let prefix_limb = {
+            let mut limb = [0u8; 8];
+            // 20th byte
+            // 0x<12 bytes of zeros>FE00..00
+            limb[3] = 0xFE;
+            u64::from_le_bytes(limb)
+        };
+        limbs[2] = prefix_limb;
+
         U256(limbs)
     };
+    const MAX_NATIVE_PRECOMPILE: U256 = {
+        // 0x<leading zeros>FE00..<len>
+        U256::from_u64(Self::NATIVE_PRECOMPILES.len() as u64)
+            .bits_or(&Self::NATIVE_PRECOMPILE_START)
+    };
 
-    // Precompile Context will be flattened to None if not calling the call_actor precompile
+    /// Precompile Context will be flattened to None if not calling the call_actor precompile.
+    /// Panics if address is not a precompile.
     pub fn call_precompile(
         system: &mut System<RT>,
         precompile_addr: U256,
         input: &[u8],
         context: PrecompileContext,
     ) -> PrecompileResult {
-        unsafe { Self::PRECOMPILES[precompile_addr.0[0] as usize - 1](system, input, context) }
+        unsafe {
+            if precompile_addr.byte(19) == NATIVE_PRECOMPILE_ADDRESS_PREFIX {
+                Self::NATIVE_PRECOMPILES[precompile_addr.0[0] as usize - 1](system, input, context)
+            } else {
+                Self::EVM_PRECOMPILES[precompile_addr.0[0] as usize - 1](system, input, context)
+            }
+        }
     }
 
+    #[inline]
+    fn is_native_precompile(addr: &U256) -> bool {
+        addr > &Self::NATIVE_PRECOMPILE_START && addr <= &Self::MAX_NATIVE_PRECOMPILE
+    }
+
+    /// Checks if word represents
     #[inline]
     pub fn is_precompile(addr: &U256) -> bool {
-        !addr.is_zero() && addr <= &Self::MAX_PRECOMPILE
+        !addr.is_zero() && (addr <= &Self::MAX_EVM_PRECOMPILE || Self::is_native_precompile(addr))
     }
```

Comment:
It would be kind of nice to unify this code a bit. We probably don't need things like "native precompile start" and friends.

1. Create some form of (internal) `lookup_precompile` function that returns an `Option<PrecompileFn<RT>>`.
2. Implement `call_precompile` in terms of that.
3. Implement `is_precompile` in terms of that.

Reply:
Then we can also drop the `bits_or` constfn.

Reply:
Specifically, the first function would be something like:

1. Check if bytes `[1, 19)` are zeros (otherwise, not a precompile).
2. Then, if the first byte is 0, it's eth if the last byte falls inside the eth precompile range.
3. If the first byte is 0xFE, it's filecoin if the last byte falls inside the filecoin precompile range.
4. Otherwise, it's not a precompile.